srml/offences/src/lib.rs

```diff
@@ -159,8 +161,111 @@ impl<T: Trait, O: Offence<T::IdentificationTuple>> ReportOffence<T::AccountId, T
 			&slash_perbill,
 		);
+
+		cache.save()
+	}
+}
+
+impl<T: Trait> Module<T> {
+	/// Compute the ID for the given report properties.
+	///
+	/// The report id depends on the offence kind, time slot and the id of offender.
+	fn report_id<O: Offence<T::IdentificationTuple>>(
+		time_slot: &O::TimeSlot,
+		offender: &T::IdentificationTuple,
+	) -> ReportIdOf<T> {
+		(O::ID, time_slot.encode(), offender).using_encoded(T::Hashing::hash)
+	}
+
+	/// Goes through all offenders in a report and record each unique offence updating indexes.
+	///
+	/// Returns `true` if there were unique reports.
+	fn collect_unique_reports<O: Offence<T::IdentificationTuple>>(
+		reporters: Vec<T::AccountId>,
+		time_slot: &O::TimeSlot,
+		offenders: Vec<T::IdentificationTuple>,
+		index_cache: &mut ReportIndexCache<T, O>,
+	) -> BTreeSet<T::IdentificationTuple> {
+		let mut new_offenders = BTreeSet::new();
+
+		for offender in offenders {
+			let report_id = Self::report_id::<O>(time_slot, &offender);
+
+			if !<Reports<T>>::exists(&report_id) {
+				new_offenders.insert(offender.clone());
+				<Reports<T>>::insert(
+					&report_id,
+					OffenceDetails {
+						offender,
+						reporters: reporters.clone(),
+					},
+				);
+
+				index_cache.insert(time_slot, report_id);
+			}
+		}
+
+		new_offenders
+	}
+}
+
+/// An auxilary struct for dealing with caches of report indexes localized for a specific offence
+/// kind.
+///
+/// It is expected that access to the indexes will be only performed through this struct while it
+/// exists.
+struct ReportIndexCache<T: Trait, O: Offence<T::IdentificationTuple>> {
+	opaque_time_slot: OpaqueTimeSlot,
+	concurrent_reports: Vec<ReportIdOf<T>>,
+	same_kind_reports: Vec<(O::TimeSlot, ReportIdOf<T>)>,
+}
+
+impl<T: Trait, O: Offence<T::IdentificationTuple>> ReportIndexCache<T, O> {
+	/// Preload indexes from the cache for the specific `time_slot` and the kind of the offence.
+	fn load(time_slot: &O::TimeSlot) -> Self {
+		let opaque_time_slot = time_slot.encode();
+
+		let same_kind_reports = <ReportsByKindIndex>::get(&O::ID);
+		let same_kind_reports =
+			Vec::<(O::TimeSlot, ReportIdOf<T>)>::decode(&mut &same_kind_reports[..])
+				.unwrap_or_default();
+
+		let concurrent_reports = <ConcurrentReportsIndex<T>>::get(&O::ID, &opaque_time_slot);
+
+		Self {
+			opaque_time_slot,
+			concurrent_reports,
+			same_kind_reports,
+		}
+	}
+
+	/// Insert a new report to the index cache.
+	fn insert(&mut self, time_slot: &O::TimeSlot, report_id: ReportIdOf<T>) {
+		// Update the list of concurrent reports.
+		self.concurrent_reports.push(report_id);
+
+		// Insert the report id into the list while maintaining the ordering by the time
+		// slot.
+		let idx = match self
+			.same_kind_reports
+			.binary_search_by_key(&time_slot, |&(ref when, _)| when)
+		{
+			Ok(idx) => idx,
+			Err(idx) => idx,
+		};
+		self.same_kind_reports.insert(idx, (time_slot.clone(), report_id));
+	}
+
+	/// Dump the indexes to the storage.
+	fn save(self) {
```

Comment:
Since you really have to call `save` at the end, maybe let's add `must_use` annotation.